pkg/controllers/jobset_controller.go

```diff
@@ -556,33 +601,41 @@ func (r *JobSetReconciler) ensureCondition(ctx context.Context, js *jobset.JobSe
 }
 
 func (r *JobSetReconciler) failJobSet(ctx context.Context, js *jobset.JobSet) error {
-	return r.ensureCondition(ctx, js, corev1.EventTypeWarning, metav1.Condition{
+	return r.ensureCondition(ctx, js, corev1.EventTypeWarning, false, metav1.Condition{
 		Type:    string(jobset.JobSetFailed),
 		Status:  metav1.ConditionStatus(corev1.ConditionTrue),
 		Reason:  "FailedJobs",
 		Message: "jobset failed due to one or more job failures",
 	})
 }
 
-func updateCondition(js *jobset.JobSet, condition metav1.Condition) bool {
+func updateCondition(js *jobset.JobSet, condition metav1.Condition, forceFalseUpdate bool) bool {
 	condition.LastTransitionTime = metav1.Now()
 	for i, val := range js.Status.Conditions {
 		if condition.Type == val.Type && condition.Status != val.Status {
 			js.Status.Conditions[i] = condition
 			// Condition found but different status so we should update
 			return true
-		} else if condition.Type == val.Type && condition.Status == val.Status {
+		} else if condition.Type == val.Type && condition.Status == val.Status && condition.Reason == val.Reason && condition.Message == val.Message {
 			// Duplicate condition so no update
 			return false
 		}
 	}
+	if forceFalseUpdate {
+		// Some conditions need an update even if false
+		// StartupPolicy is one example.
```

Comment:
In the comment can you elaborate on why this forced false update is necessary for startupPolicy? It's not immediately clear to me.

Reply:
updated.

added a comment.

```
		// Some conditions need an update even if false
		// StartupPolicy is one example
		// If startup policy is not specified, then
		// we assume that there was no startup policy ever applied
		// We use the false condition to signify progress of StartupPolicy.
```